File: 2025/SoyoungJung/week4/2025-06-29-Binary-Search.ts
function binarySearch(nums: number[], target: number): number {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    const guess = nums[mid];

    if (guess === target) {
      return mid;
    } else if (guess < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return -1;
}

File: 2025/shinyoung/2025-06-29binary-search.py
def search(nums, target):
    def dfs(left, right):
        middle = left + (right - left) // 2
        if right-left < 1:
            if nums[left] == target:
                return middle
            else:
                return -1
        if target == nums[middle]:
            return middle
        if target < nums[middle]:
            return dfs(left, middle-1)
        if target > nums[middle]:
            return dfs(middle+1, right)
    return dfs(0, len(nums)-1)


# print(search([-1, 0, 3, 5, 9, 12], target=9))
# print(search([-1, 0, 3, 5, 9, 12], target=2))
print(search([5], target=5))
print(search([2, 5], target=0))
